### PR TITLE
added txpool methods

### PR DESCRIPTION
**Pull request by dankostiuk**
# Description

Added the 3 geth RPC endpoints for `txpool` to support `txpool_content`, `txpool_inspect`, and `txpool_status` (see https://geth.ethereum.org/docs/rpc/ns-txpool).

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: jsonrpc/txpool_endpoint.go
Comment (Kourin1996) [inline jsonrpc/txpool_endpoint.go]: I suppose we can't use `toTransaction` here because tx is not in any block yet and `toTransaction` probably will throw panic. We need to use different function and set "0x00...." to block hash and nil to block number in the response according to docs.

File: jsonrpc/txpool_endpoint.go
Comment (Kourin1996) [inline jsonrpc/txpool_endpoint.go]: To convert uint64 to string, you can use `strconv.FormatUint` which is more efficient function.

```
key := strconv.FormatUint(nonce, 10)
```

File: jsonrpc/txpool_endpoint.go
Comment (dankostiuk) [inline jsonrpc/txpool_endpoint.go]: blockNumber and transactionIndex must be included and nil according to the geth spec.

File: jsonrpc/txpool_endpoint.go
Comment (dankostiuk) [inline jsonrpc/txpool_endpoint.go]: Here if R or S is `nil`, argBytes will format the response json as `0x` however this isn't an acceptable value. Therefore, I initialized a 0 byte array if either of these fields are `nil` so that the response has values to the right of x, for example `0x0`.

File: jsonrpc/txpool_endpoint.go
Comment (dankostiuk) [reply jsonrpc/txpool_endpoint.go]: Thanks @Kourin1996 , I added that and will keep that in mind for any future changes.

File: txpool/txpool.go
Comment (Kourin1996) [inline txpool/txpool.go]: You might want to create new map outside second loop (between line 210 and 211). In current implementation, result shows only one transaction per an account. (According to doc, you can see transactions by multiple nonce if exist)

File: txpool/txpool.go
Comment (Kourin1996) [inline txpool/txpool.go]: This is the similar one. It's better to create new map only if not exist.

File: jsonrpc/txpool_endpoint.go
Comment (Kourin1996) [inline jsonrpc/txpool_endpoint.go]: You don't need to put V & R & S in the response according to doc https://geth.ethereum.org/docs/rpc/ns-txpool

**Review by Kourin1996 — CHANGES_REQUESTED** [review]
Hello. Thank you for your PRs. I noticed some points and put comment. It'd be great if you would check it.

File: jsonrpc/txpool_endpoint.go
Comment (zivkovicmilos) [inline jsonrpc/txpool_endpoint.go]: I'm not sure about the naming on this. `gethTransaction` implies that the transaction is conforming to specific geth services, when in fact it is just emulating the same functionality.

File: jsonrpc/txpool_endpoint.go
Comment (zivkovicmilos) [reply jsonrpc/txpool_endpoint.go]: Thank you for the explanation. Could you add it to the code as well, since these comments will not be visible in the future when someone is going over it?

File: jsonrpc/txpool_endpoint.go
Comment (zivkovicmilos) [inline jsonrpc/txpool_endpoint.go]: This is probably a nitpick, but throughout the entire project we've never used comments like this. Please place multiline comments in // instead of /* */. Function comments should have a short description as to what they do, and you can additionaly link to the spec after the explanation

File: jsonrpc/txpool_endpoint.go
Comment (zivkovicmilos) [inline jsonrpc/txpool_endpoint.go]: Same as with the previous comment, please conform it to what is present in the codebase

File: jsonrpc/txpool_endpoint.go
Comment (zivkovicmilos) [inline jsonrpc/txpool_endpoint.go]: Same as with the previous 2 comments, please conform this

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Overall it looks good.

Please review the comments I've left, and it will be 💯

File: txpool/txpool.go
Comment (dankostiuk) [reply txpool/txpool.go]: Thanks @Kourin1996 , for both these comments, I created a new map only if one at the particular address doesn't exist. I found out that the scenario exists where a particular address might not have any txs, so this approach ensures there's always at least 1 tx per address.

**Review by Kourin1996 — APPROVED** [review]
Looks good to me 👍 Thank you for new features.

**Review by lazartravica — APPROVED** [review]
LGTM! I especially liked how you wrote tests both for the txpool and txpool_endpoint in different places 💯

**Review by Kourin1996 — APPROVED** [review]

File: txpool/txpool.go
Comment (Vuksan) [inline txpool/txpool.go]: You are returning `queuedTxs, pendingTxs` but in `txpool_endpoint.go` when you call this method you are assigning them in a different order: 
```
pendingTxs, queuedTxs := t.d.store.GetTxs()
```

File: txpool/txpool_test.go
Comment (Vuksan) [inline txpool/txpool_test.go]: Shouldn't there be 3 queued transactions (`txn1, txn2 and txn3`)?

**Review by Vuksan — CHANGES_REQUESTED** [review]
Thank you for adding these methods 🙇 
Can you please take a look at the comments I left, I think you made a small mistake?

File: txpool/txpool.go
Comment (dankostiuk) [reply txpool/txpool.go]: Thanks @Vuksan! Fixed this. The return is now pending followed by queued txs in all occurrences exactly how is done with geth (https://github.com/ethereum/go-ethereum/blob/d3e3a460ec947c9e1e963d1a35f887d95f23f99d/core/tx_pool.go#L482)

File: txpool/txpool_test.go
Comment (dankostiuk) [reply txpool/txpool_test.go]: Yes, however in this test 2 txs were sent from 1 account. I now separated the `from` of each tx so that the length of the queued map is 3 which makes it easier to follow.

**Comment by ferranbt** [comment]
I do not think these methods should be here. Polygon-sdk has two main interfaces: JsonRPC and GRPC. The first one is a generic interface to send Ethereum transactions, think of the minimum methods required to be usable with a client like etherjs. Thus, it was not intended to implement all the Geth namespace.

On the other side, the GRPC interface was meant to be the place to add these utility methods, besides, there is already a TxPool namespace there.

**Comment by mrwillis** [comment]
@ferranbt I see your point of wanting to make the sdk RPC interface as generic as possible but I think these client-specific methods (also this one https://github.com/0xPolygon/polygon-sdk/issues/118) add a lot of value to the project.

We've created this PR to make polygon-sdk compatible with many block explorers as I don't believe the support of gRPC on block explorers in the ethereum ecosystem is strong. If the SDK can conform to a geth or parity interface, polygon-sdk can out of the box work with many explorers and node providers and not require teams to implement them themselves.

I'd also note that clients like ethers.js use these methods in many cases to offer additional functionality. 

Maybe these can be exposed via a flag similar to how geth offers you options to expose different namespaces?

**Comment by lazartravica** [comment]
> I do not think these methods should be here. Polygon-sdk has two main interfaces: JsonRPC and GRPC. The first one is a generic interface to send Ethereum transactions, think of the minimum methods required to be usable with a client like etherjs. Thus, it was not intended to implement all the Geth namespace.
> 
> On the other side, the GRPC interface was meant to be the place to add these utility methods, besides, there is already a TxPool namespace there.

Thank you @ferranbt for keeping track of the pull requests 🔥 

I must say I agree with @mrwillis on this one, as we want to support people connecting polygon-sdk to blockexplorers. On the other side I agree with you that indiscriminately adding endpoints exposed in other clients to the polygon-sdk is not something that we should do, but rather keep the set of endpoints to a functional minimum.

I guess the only point where we have a conflict regarding this is whether querying the txpool should be a part of the minimal set of JSON-RPC methods or not, or is there a deeper technical issue that you see here?

**Review by Vuksan — APPROVED** [review]
From a technical perspective, you have my approval.
Thank you for the contribution and for making the requested changes.

As for the strategic decision, let's wait for @ferranbt to answer @lazartravica's comment.

**Comment by ferranbt** [comment]
The only technical issue is to maintain the same sets of methods in two places (JSON-RPC and GRPC) and how much will grow the interface between the JSON-RPC layer.

For me, the minimal set of JSON-RPC methods should be the ones implemented in Infura since that is the main gateway used by Ethereum compatible end users. I understand that having block exporters is important but not sure at which point you draw the line.

**Comment by Vuksan** [comment]
This is how Celo did it: https://github.com/celo-org/celo-blockchain/blob/master/internal/web3ext/web3ext.go#L664-L688

**Comment by ferranbt** [comment]
There is something else I forgot. If you include debug or txpool namespace directly on JSON-RPC you will likely have to add support to enable different namespaces for each transport (i.e. rpcapi and wsapi in Geth) since you are mixing endpoints for two different stakeholders.

A different approach would be to implement these methods on the GRPC layer only and then have a separate microservice/repo that translates JSON-RPC calls to the GRPC endpoints.

**Comment by lazartravica** [comment]
@ferranbt Thank you for the comments! Since this is a community opened PR, I will close the pull request and merge, so that we can redesign the way we expose non-essential APIs  internally.

**Pull request merged**
